Provide a latest-beta extension download link (bug 539318)

palant opened the pull request:
Sorry about reviving bug 539318 but this feature is actually important to us. We need to link to the latest file in the beta channel directly in order to simplify installation both for our users and third-party testers.

magopian commented:
I have a bad feeling about this PR, just look at its number ;)

Thanks for this work, I'm going to review it asap.

magopian commented on apps/versions/tests.py:
```diff
@@ -833,6 +840,22 @@ def test_success(self):
+        eq_(r.status_code, 302)
+        beta_file_url = reverse('downloads.file', args=[self.beta_file.id])
+        url = beta_file_url + '/' + self.beta_file.filename
+        assert r['Location'].endswith(url), r['Location']
```
You don't need the `, r['Location']`, because we're using [pytest](http://pytest.org/latest/assert.html?highlight=introspection) and its clever "assert introspection", which gives you all the needed details if a test fails, for free!

palant replied:
I have exactly zero idea about the testing framework, I merely copied what other tests were doing :-)

Sure, will do things "the good way."

magopian commented on apps/versions/tests.py:
```diff
@@ -833,6 +840,22 @@ def test_success(self):
+        eq_(r.status_code, 302)
+        beta_file_url = reverse('downloads.file', args=[self.beta_file.id])
+        url = self.file_url + '/' + self.file.filename
+        assert r['Location'].endswith(url), r['Location']
```
Same here, no need for the extra message part on the assert

magopian commented on apps/versions/tests.py:
```diff
@@ -833,6 +840,22 @@ def test_success(self):
 
+    def test_beta(self):
+        r = self.client.get(self.latest_beta_url)
+        eq_(r.status_code, 302)
```
Since we switched to pytest, we don't use `eq_` anymore, but instead `assert`:

```assert r.status_code == 302```

magopian commented on apps/versions/tests.py:
```diff
@@ -833,6 +840,22 @@ def test_success(self):
         self.assert_served_by_mirror(self.client.get(self.latest_url))
 
+    def test_beta(self):
+        r = self.client.get(self.latest_beta_url)
```
I know you've seen quite a few tests like that (using one letter variables), but please don't do that. We try to write new tests "the good way", not using one letter variables.

Eg: ```response = self.client.get(self.latest_beta_url)```

magopian commented on apps/versions/tests.py:
```diff
@@ -833,6 +840,22 @@ def test_success(self):
+        beta_file_url = reverse('downloads.file', args=[self.beta_file.id])
+        url = self.file_url + '/' + self.file.filename
+        assert r['Location'].endswith(url), r['Location']
+
```
I would love a test here for the case where the addon is public, but doesn't have a beta file (and thus returns a 404? From the current code, I believe it'll return the latest version, but I'm not sure that's what we want, see my comment below).

magopian replied:
Btw, I wouldn't mind if you split the test to have one small test for each of the three cases:
1/ public, with a beta version
2/ public, no beta version
3/ not public (with or without a beta version, we don't really care, we should never be in a state where a non public addon have a beta version, as beta uploads are only allowed on public addons)

magopian commented on apps/versions/views.py:
```diff
@@ -110,12 +110,16 @@ def guard():
     platforms = [amo.PLATFORM_ALL.id]
     if platform is not None and int(platform) in amo.PLATFORMS:
         platforms.append(int(platform))
+    if beta and addon.show_beta:
```
What happens if there's a `-beta` in the url, but `show_beta` is False (because for example there's no beta)? We'll end up serving a version that isn't beta, right? Shouldn't we instead return a 404?

palant replied:
Yes, currently we would server a stable release. I was wondering whether providing something would be better than giving nothing in that case. Then again, it would probably be unexpected behavior - and it's not like the beta channel would ever go away once it is present. So I guess 404 would indeed be better.

magopian commented:
This is excellent, thanks a lot for taking the time to contribute this patch!

On top of the comments I added, would it make sense to have this url publicly available somewhere in the public pages?

palant commented:
I've addressed your comments. As to showing the URL - it would make sense to use it in the beta channel section but this isn't entirely trivial so I'm still looking into it.

palant commented:
Done, now that link is actually being used. However, I didn't test the change to `apps/addons/templates/addons/details_box.html` - from what I can tell, that's just dead code there. This template is used for the review page but the beta channel is behind the `show_actions` flag which is never being set.

magopian commented on apps/addons/buttons.py:
```diff
@@ -90,7 +90,11 @@ def __init__(self, addon, app, lang, version=None, show_contrib=True,
+        if version == 'beta':
+            self.version = addon.current_beta_version
+        else:
+            self.version = version or addon.current_version
```
Mmm, this works, but I'd rather not have a parameter that accepts either a version, `None` or `'beta'`. How about adding a new `beta=False` parameter, and then:
```
self.beta = beta
if self.beta:
    self.version = addon.current_beta_version
else:
    self.version = version or addon.current_version
```

You would then call the `install_button` helpers below with `version=None, beta=True`.

What do you think?

palant replied:
I had a hunch that you wouldn't like it :-)

Heh, will have to do this properly I guess - will make the changes a bit more complicated again.

magopian commented on apps/addons/buttons.py:
```diff
@@ -157,6 +161,8 @@ def file_details(self, file):
         if self.latest and (
                 self.addon.status == file.status == amo.STATUS_PUBLIC):
             url = file.latest_xpi_url()
+        elif self.latest_beta and self.addon.show_beta:
```
This would then be changed to `elif self.beta and self.addon.show_beta:`

magopian commented:
This is looking better, congrats on navigating the code to find your way, I know it's complex. And the install buttons are one of the trickiest places to modify without introducing new bugs (so let's cross fingers!).

I'd love a test for the new `beta` parameter of the `InstallButton`, in `apps/addons/tests/test_buttons.py` I believe.

palant commented:
There we go, added an `is_beta` parameter - there is also an `is_beta` property so I wanted to keep things consistent. They would be more consistent if one could assume that `addon.current_version.is_beta` would always be `False`, but it isn't, at least not in the unit tests. Maybe the unit tests are bogus but this isn't something I will be able to figure out today - and neither adding my own tests for that parameter.

magopian commented:
Ah, I didn't realise there was already a `is_beta` property there... which makes it all very confusing, right?
As I said, this code is very brittle, technical debt ridden, and hard to maintain... it's high on my list of things to refactor.
How about naming this parameter `latest_beta` instead? It clarifies the fact that we want to display the latest beta, instead of saying "we want to display a beta file, and as the version is None, then it's the latest beta" or whatever ...

Also please add a docstring explaining at least the use of this new parameter (there should have been one already...).

Thanks!

palant commented:
Ok, the parameter is called `latest_beta` now. It's somewhat less confusing but it also created an invalid scenario where both `version` and `latest_beta` are set. I thought I'd rather assert for it rather than dropping that condition silently. Let me know if you can think of a better way (also, maybe that assert should be in the `InstallButton` class).

magopian commented on apps/addons/buttons.py:
```diff
@@ -16,8 +16,16 @@
+    """
+    assert not version or not latest_beta, ('version and latest_beta '
+                                            'parameters cannot be specified '
+                                            'at the same time')
```
This is fine. I would have written it `assert not (version and latest_beta)`, what do you think?

magopian commented:
This is all very good, thank you! I'm going to squash your commits, and merge them, with the [small last nit I have](https://github.com/mozilla/olympia/pull/666#discussion_r37619116).

Thanks!

magopian commented:
Fixed by 4cac6e61968d5d229156b92af07dbbaa2e769658, thanks again @palant !
